app/services/usps_in_person_proofing/proofer.rb

```diff
@@ -102,17 +102,28 @@ def request_enrollment_code(unique_id)
     end
 
     # Makes a request to retrieve a new OAuth token
-    # and modifies self to store the token and when
+    # and updates the cached token value
     # it expires (15 minutes).
-    # @return [String] the token
+    # @return [String] Auth token
     def retrieve_token!
-      body = request_token
-      @@token_expires_at = Time.zone.now + body['expires_in']
-      @@token = "#{body['token_type']} #{body['access_token']}"
+      token, expires_in, token_type = request_token.fetch_values(
+        'access_token',
+        'expires_in',
+        'token_type',
+      )
+      authz = "#{token_type} #{token}"
+      Rails.cache.write(
+        API_TOKEN_CACHE_KEY, authz,
+        expires_at: Time.zone.now + expires_in
+      )
+      authz
     end
 
-    def token_valid?
-      token.present? && token_expires_at.present? && token_expires_at.future?
+    # Checks the cache for an unexpired token and returns it. If the cache has expired, retrieves
+    # a new token and returns it
+    # @return [String] Auth token
+    def token
+      Rails.cache.read(API_TOKEN_CACHE_KEY) || retrieve_token!
     end
```

Comment:
Does the Rails cache include an in-memory cache or does it go straight to Redis? If it's the latter, then I would suggest keeping the class/module field.

e.g. in the use case of a large volume in the the USPS proofing job, this could mean hundreds of unnecessary trips to Redis.

Reply:
You would still need to be mindful of the TTL though.

Reply:
My understanding is that it was the primary _goal_ of this work to have it be stored in a Redis cache, so that the same USPS token would be reused across IdP instances.

The Redis cost is definitely an interesting factor I hadn't given much thought to previously, but then we'd still want to weigh it against the additional cost of roundtrips to USPS for IdP-instance-specific tokens.

Reply:
The intention would be to use the same token across instances, but that token could be cached in-memory.

Reply:
Ah, I think I see what you're saying now -- still use Redis, but also having a local copy to avoid repeatedly calling to Redis while the token is assumed to be valid? (what @sheldon-b was referring to in "two layers of caching")

Sounds like it could be a good idea, but also likely to be tricky to implement 😅

Reply:
That's what I'm talking about.

Pseudocode would look like:

```
If local cache exists and TTL has not passed:
  Use local version
Else:
  Get Redis value

  If Redis value is missing:
    Authenticate w/ USPS
    SET Redis value w/ NX and PX options
  
  Get Redis value  
  Locally cache and use Redis value
```

- [Redis SET](https://redis.io/commands/set/)

The TTL should be identical between the local cache and Redis, but both should be very slightly before the real expiry of the token.